File: DataStructure/LinkedList/LinkedList.py
class Node:
    def __init__( self, num ):
        self.data = num
        self.next = None

class LinkedList:
    def __init__( self ):
        self.root = None
        self.size = 0

    def add( self, num ):
        newNode = Node( num )
        newNode.next = self.root
        self.root = newNode
        print( "Added:" , self.root.data ) 
        self.size += 1

    # Add appropriate message when the list is empty
    def delete( self ):
        print( "Deleted:" , self.root.data ) 
        self.root = self.root.next
        self.size -= 1

    def elementAt( self, index ):
    #     LinkedList: 10 <- 20 <- 30 <- 40
    #     Index:      0     1     2     3
    
        length = self.size-1
        temp = self.root

        while( index != length ):
            temp = temp.next
            length -= 1   

        return temp

    def print( self ) :
        if(self.root!=None):
            temp = self.root
            print( "LinkedList Values: ", end="" )
            while( temp != None ):
                print( temp.data , end=" ")
                temp = temp.next
            print()
        else:
            print("Linked List is cleared")


    
    def printReverse( self ):
        node = self.root
        values = []
        while node:
            values.append(node.data)
            node = node.next

        for value in reversed(values):
            print(value, end=" ")
        print()


    def reverse( self ): 
        prev = None
        current = self.root 
        while(current is not None): 
            next = current.next
            current.next = prev 
            prev = current 
            current = next
        self.root = prev
    def sort( self ):
        current = self.root;  
        index = None;  
          
        if(self.root == None):  
            return;  
        else:  
            while(current != None):  
                #Node index will point to node next to current  
                index = current.next;  
                  
                while(index != None):  
                    #If current node's data is greater than index's node data, swap the data between them  
                    if(current.data > index.data):  
                        temp = current.data;  
                        current.data = index.data;  
                        index.data = temp;  
                    index = index.next;  
                current = current.next;  

    def clear( self ) :
        self.root=None
        self.size=0

    def contains( self, value ) :
        temp = self.root
        while( temp != None ):
            if(temp.data==value):
                print(value,"exist in linked list")
                return
            temp=temp.next
        print(value,"Not exist in linked list")
        

ll = LinkedList()
ll.add( 1 )
ll.print()

ll.add( 2 )
ll.print()

ll.delete()
ll.print()

ll.add(3)
ll.add(7)
ll.add(5)
ll.add(9)
ll.print()

ind = 1 
node = ll.elementAt( ind )
print( "Element at index" , ind , ":" , node.data )

print("linked list printing in reverse order: ",end="")
ll.printReverse()

ll.contains(15)
ll.contains(7)

ll.sort()
print("after sorting the elements in linked list")
ll.print()

ll.reverse()
print("Aftrer reversing the elements in linked list")
ll.print()
ll.clear()
print("After clearing elements in linked list")
ll.print()

File: DataStructure/LinkedList/LinkedList.js

class Node {
    constructor( num ) {
        this.data = num
        this.next = null
    }
}
class LinkedList {
    constructor() {
        this.root = null
        this.size = 0
    }

    add( num ) {
        const newNode = new Node( num )
        newNode.next = this.root
        this.root = newNode
        console.log( "Added:" , this.root.data ) 
        this.size += 1
    }

    delete() {
        console.log( "Deleted:" , this.root.data ) 
        this.root = this.root.next
        this.size -= 1
    }

    elementAt( index ) {
        //      LinkedList: 10 <- 20 <- 30 <- 40
        //      Index:      0     1     2     3
        
        let length = this.size-1
        let temp = this.root

        while( index != length ) {
            temp = temp.next
            length -= 1   
        }

        return temp
    }

    // Add appropriate message when the list is empty
    print() {
        let temp = this.root
        const ele = []
        while( temp != null ) {
            ele.push( temp.data )
            temp = temp.next
        }
        console.log( "LinkedList Values:", ele )
    }

    // Test and Complete the following methods
    printReverse( ) {
    }

    reverse() {
        let prev,next;
        let current = this.root ;
        while(current){ 
            next = current.next
            current.next = prev 
            prev = current 
            current = next
            }
        this.root = prev;
    }
    sort() {
        let current = this.root;  
        let index,temp; 
          
        if(!this.root)  
            return;  
        else{
            while(current){ 
                index = current.next;  
                  
                while(index){ 
                	if(current.data > index.data){  
                        temp = current.data;  
                        current.data = index.data;  
                        index.data = temp; 
                    }
                    index = index.next;
                }
                current = current.next;
            }
            }
    }

    clear() {
        this.root=null;
        this.size=0;
        console.log("Linked List is Cleared");
    }

    contains(value) {
        let temp=this.root;
        while(temp){
        if(temp.data==value){
        	console.log("Linked List contains "+value);
            return;
        }
        temp=temp.next;
        }
        console.log("Linked List doesn't contain "+value);
    }
    
}

ll = new LinkedList()
ll.add( 1 )
ll.print()

ll.add( 2 )
ll.print()

ll.delete()
ll.print()

ll.add(3)
ll.add(4)
ll.add(2)
ll.print()

ind = 1 
node = ll.elementAt( ind )
console.log( "Element at index" , ind , ":" , node.data ) 

ll.contains(3)

ll.reverse();
console.log("After reversing element");
ll.print();

ll.sort();
console.log("After sorting element");
ll.print();

ll.clear();
